numpy/distutils/system_info.py

```diff
@@ -919,44 +945,50 @@ def _find_lib(self, lib_dir, lib, exts):
             for prefix in lib_prefixes:
                 p = self.combine_paths(lib_dir, prefix + lib + ext)
                 if p:
-                    break
+                    assert len(p) == 1
+                    # p[0] is the full path to the binary library file.
+                    if filefilter is None or filefilter(p[0]):
+                        break
+                    else:
+                        p = []
             if p:
-                assert len(p) == 1
                 # ??? splitext on p[0] would do this for cygwin
                 # doesn't seem correct
                 if ext == '.dll.a':
                     lib += '.dll'
                 if ext == '.lib':
                     lib = prefix + lib
                 return lib
-
         return False
 
-    def _find_libs(self, lib_dirs, libs, exts):
+    def _find_libs(self, lib_dirs, libs, exts, filefilter=None):
         # make sure we preserve the order of libs, as it can be important
         found_dirs, found_libs = [], []
         for lib in libs:
             for lib_dir in lib_dirs:
-                found_lib = self._find_lib(lib_dir, lib, exts)
+                found_lib = self._find_lib(lib_dir, lib, exts,
+                                           filefilter=filefilter)
                 if found_lib:
                     found_libs.append(found_lib)
                     if lib_dir not in found_dirs:
                         found_dirs.append(lib_dir)
                     break
         return found_dirs, found_libs
 
-    def _check_libs(self, lib_dirs, libs, opt_libs, exts):
+    def _check_libs(self, lib_dirs, libs, opt_libs, exts, filefilter=None):
         """Find mandatory and optional libs in expected paths.
 
         Missing optional libraries are silently forgotten.
         """
         if not is_sequence(lib_dirs):
             lib_dirs = [lib_dirs]
         # First, try to find the mandatory libraries
-        found_dirs, found_libs = self._find_libs(lib_dirs, libs, exts)
+        found_dirs, found_libs = self._find_libs(lib_dirs, libs, exts,
+                                                 filefilter=filefilter)
         if len(found_libs) > 0 and len(found_libs) == len(libs):
             # Now, check for optional libraries
-            opt_found_dirs, opt_found_libs = self._find_libs(lib_dirs, opt_libs, exts)
+            opt_found_dirs, opt_found_libs = self._find_libs(
+                lib_dirs, opt_libs, exts, filefilter=filefilter)
```

Comment:
The linker picks the first library file with the correct name in lib_dirs it finds.
Does the filefilter logic here match that, i.e., stop at the first file (and then accept/reject the whole shebang)?

Reply:
If the filter returns False, `check_libs` will act like the file doesn't exist.  If there is some combination of inputs that would make the linker find that file during the build, then that could be a problem.   I think I see what you're worried about, but to be clear, could you describe a concrete example where something like this might happen?

Reply:
For example:
```
/usr/lib/libblas.so -> Accelerate
/home/mypath/libblas.so -> Some other BLAS
```
With library_dirs `/usr/lib:/home/mypath` I guess the question is, which library directories get passed to the linker?

Reply:
I'm experimenting with something like that now.  What is the standard way to tell numpy to use blas, cblas and lapack from a custom directory?  Is site.cfg only for known libraries (i.e. [mkl], [atlas], etc), or can I add a section such as like [mylib]?  (My first attempt failed, but I had some other issues, so I'm still experimenting.)

Reply:
Nevermind. `site.cfg` like this seems to do it:

```
[blas]
library_dirs = /Users/warren/foo/lib
include_dirs = /Users/warren/foo/include
libraries = refblas,cblas

[lapack]
library_dirs = /Users/warren/foo/lib
include_dirs = /Users/warren/foo/include
libraries = lapack
```

Reply:
Since the code in numpy.distutils is attempting to do what the linker will also eventually do (i.e. find the binary library file), is it feasible to replace the use of -l and -L for a given library with the full path to the binary itself?  So instead of `-lblas -L/home/mypath`, the linker would be given the filename itself `/home/mypath/libblas.so` (or whatever file was found, whether that be .so, .a, .dylib, etc.).